### PR TITLE
[PATCH API-NEXT v4] Another IPsec API example

### PR DESCRIPTION
**Pull request by lumag**

### CODE REVIEW

**Comment by muvarov** [comment]
In API we defined one pktio with name "loop" as loopback device supported everywhere. 

```
 * @note The device name "loop" is a reserved name for a loopback device used
 *       for testing purposes.
```
There was a talk to support multiple loop devices and there wasn't agreement to do that. From other side you do not break API but define just additional loop devices for linux-generic. That is ok. But example will not run on platform which can not pktio_open("loopX").  A pktio also can be switched to loop mode with odp_pktio_config_t->enable_loop.  So in general other platform can redefine which pktios they want use for loop. Like export ODP_PKTIO_LOOP1="eth1" ODP_PKTIO_LOOP2="eth2". In that case it will be more general solution.

**Comment by lumag** [comment]
On the rest of platforms it is possible to use plain devices without attached stream generator. Like in the existing ipsec exaple.

File: example/ipsec_api/odp_ipsec.c [hunk 1]
```diff
@@ -0,0 +1,1248 @@
...
+		return PKT_POSTED;
+	} else {
+		int out = 1;
+
+		rc = odp_ipsec_in(ppkt, 1, ppkt, &out, &in_param);
+		if (rc <= 0)
+			return PKT_DROP;
+
+		return PKT_CONTINUE;
+	}
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec.c]: Checkpatch complains about this construct. Prefers no else after return:
```
WARNING: else is not generally useful after a break or return
#795: FILE: example/ipsec_api/odp_ipsec.c:603:
+		return PKT_POSTED;
+	} else {
```

File: example/ipsec_api/odp_ipsec.c [hunk 2]
```diff
@@ -0,0 +1,1248 @@
...
+		return PKT_POSTED;
+	} else {
+		int out = 1;
+
+		rc = odp_ipsec_out(ppkt, 1, ppkt, &out, &out_param);
+		if (rc <= 0)
+			return PKT_DROP;
+
+		return PKT_CONTINUE;
+	}
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec.c]: Similar checkpatch issue here:
```
WARNING: else is not generally useful after a break or return
#859: FILE: example/ipsec_api/odp_ipsec.c:667:
+		return PKT_POSTED;
+	} else {
```

File: example/ipsec_api/odp_ipsec_sa_db.c [hunk 3]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_sa_db.c
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_sa_db.c]: Checkpatch doesn't like this construct:
```
WARNING: adding a line without newline at end of file
#1751: FILE: example/ipsec_api/odp_ipsec_fwd_db.c:1:
+../ipsec/odp_ipsec_fwd_db.c
```
Missing newline?

File: example/ipsec_api/odp_ipsec_fwd_db.h [hunk 4]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_fwd_db.h
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_fwd_db.h]: Similar issues here. Missing newline?
```
WARNING: adding a line without newline at end of file
#1759: FILE: example/ipsec_api/odp_ipsec_fwd_db.h:1:
+../ipsec/odp_ipsec_fwd_db.h
```

File: example/ipsec_api/odp_ipsec_misc.h [hunk 5]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_misc.h
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_misc.h]: Also here:
```
WARNING: adding a line without newline at end of file
#1767: FILE: example/ipsec_api/odp_ipsec_misc.h:1:
+../ipsec/odp_ipsec_misc.h
```

File: example/ipsec_api/odp_ipsec_run_ah_in [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+# Test input AH
+#  - 2 loop interfaces
+#  - 10 packets
+#  - Specify API mode on command line
+odp_ipsec -i loop1,loop2 \
+-r 192.168.111.2/32:loop1:08.00.27.76.B5.E0 \
+-p 192.168.222.0/24:192.168.111.0/24:in:ah \
+-a 192.168.222.2:192.168.111.2:md5:300:27f6d123d7077b361662fc6e451f65d8 \
+-s 192.168.222.2:192.168.111.2:loop2:loop1:10:100 \
+-c 2 "$@"
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_run_ah_in]: Checkpatch seems confused here. Is it expecting a .sh suffix? Any thoughts on this @muvarov?
```
ERROR: do not set execute permissions for source files
#1770: FILE: example/ipsec_api/odp_ipsec_run_ah_in
```
The rest of the shell files are similarly flagged.

File: example/ipsec_api/odp_ipsec_stream.h [hunk 7]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_stream.h
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_stream.h]: Missing newline here too.
```
WARNING: adding a line without newline at end of file
#2020: FILE: example/ipsec_api/odp_ipsec_stream.h:1:
+../ipsec/odp_ipsec_stream.h
```

File: example/ipsec_api/odp_ipsec_stream.c [hunk 8]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_stream.c
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_stream.c]:
```
WARNING: adding a line without newline at end of file
#2012: FILE: example/ipsec_api/odp_ipsec_stream.c:1:
+../ipsec/odp_ipsec_stream.c
```

File: example/ipsec_api/odp_ipsec_sp_db.h [hunk 9]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_sp_db.h
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_sp_db.h]:
```
WARNING: adding a line without newline at end of file
#2004: FILE: example/ipsec_api/odp_ipsec_sp_db.h:1:
+../ipsec/odp_ipsec_sp_db.h
```

File: example/ipsec_api/odp_ipsec_sp_db.c [hunk 10]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_sp_db.c
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_sp_db.c]:
```
WARNING: adding a line without newline at end of file
#1996: FILE: example/ipsec_api/odp_ipsec_sp_db.c:1:
+../ipsec/odp_ipsec_sp_db.c
```

File: example/ipsec_api/odp_ipsec_sa_db.h [hunk 11]
```diff
@@ -0,0 +1 @@
+../ipsec/odp_ipsec_sa_db.h
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_api/odp_ipsec_sa_db.h]:
```
WARNING: adding a line without newline at end of file
#1988: FILE: example/ipsec_api/odp_ipsec_sa_db.h:1:
+../ipsec/odp_ipsec_sa_db.h
```

File: example/ipsec_api/odp_ipsec.c (hunk 1 above)
Comment (lumag) [reply example/ipsec_api/odp_ipsec.c]: And I prefer current code. It is cleaner from my point of view.
```c
if (!SYNC) {
  odp_ipsec_in_enq();
  return PKT_POSTED:
} else {
  odp_ipsec_in();
  return PKT_CONTINUE;
}
```

File: example/ipsec_api/odp_ipsec_fwd_db.h (hunk 4 above)
Comment (lumag) [reply example/ipsec_api/odp_ipsec_fwd_db.h]: Ugh. This (and several following warnings) are barfing on a symlink. There should be no newline at such files.

File: example/ipsec_api/odp_ipsec_run_ah_in (hunk 6 above)
Comment (lumag) [reply example/ipsec_api/odp_ipsec_run_ah_in]: @Bill-Fischofer-Linaro yes, it expects .sh files. I can change filenames, if you wish.

Comment (Bill-Fischofer-Linaro) [reply example/ipsec_api/odp_ipsec_run_ah_in]: We use that convention elsewhere (_e.g.,_ in the various `test/performance` scripts) so I think that's a good convention to follow.

**Comment by lumag** [comment]
@muvarov @Bill-Fischofer-Linaro updated:
- renamed scripts ot include `.sh` suffix
- cleaned up Makefile.am

**Comment by muvarov** [comment]
Merged.

**Comment by lumag** [comment]
@muvarov you have merged #339, but closed #340. Could you please fix that?

**Comment by lumag** [comment]
@muvarov @Bill-Fischofer-Linaro could you please reopen this PR. It was not merged.

**Comment by Bill-Fischofer-Linaro** [comment]
Reopened per @lumag request.

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**